test/runtests.jl

```diff
@@ -256,3 +256,103 @@ end
+    # Set station coordinates
+    ist = rand(1:model_params.nx, model_params.nobs)
+    jst = rand(1:model_params.ny, model_params.nobs)
+    stations = (ist = ist, jst = jst)
+    cov_ext = ParticleDA.extended_covariance(0.0, 0.5 * model_params.y_length, model_params, filter_params)
+    @test cov_ext ≈ exp(-0.5 * model_params.y_length / (2 * filter_params.lambda_cov))
+    @test cov_ext ≈ ParticleDA.extended_covariance(2.0 * model_params.x_length, 0.5 * model_params.y_length, model_params, filter_params)
+    @test cov_ext ≈ ParticleDA.extended_covariance(0.0, 1.5 * model_params.y_length, model_params, filter_params)
+    arr = rand(ComplexF64,10,10)
+    arr2 = zeros(ComplexF64,10,10)
+    arr3 = zeros(ComplexF64,10,10)
+    ParticleDA.normalized_2d_fft!(arr2,arr,model_params)
+    ParticleDA.normalized_inverse_2d_fft!(arr3,arr2,model_params)
+    @test arr ≈ arr3
+
+    cov_1 = zeros(model_params.nobs,model_params.nx_ext * model_params.ny_ext)
+    cov_2 = zeros(model_params.nx * model_params.ny, model_params.nobs)
+    cov_3 = zeros(model_params.nobs,model_params.nobs)
+    ParticleDA.covariance_stations_extended_grid!(cov_1,model_params,filter_params,stations)
+    ParticleDA.covariance_stations_grid!(cov_2,model_params,filter_params,stations)
+    ParticleDA.covariance_stations!(cov_3,model_params,filter_params,stations)
+    @test all(isfinite.(cov_1))
+    @test all(isfinite.(cov_2))
+    @test all(isfinite.(cov_3))
+    @test cov_3 == Symmetric(cov_3)
+
+    height = rand(model_params.nx, model_params.ny, filter_params.nprt)
+    obs = randn(model_params.nobs)
+    mat_off = ParticleDA.init_offline_matrices(model_params, filter_params, stations)
+    mat_on = ParticleDA.init_online_matrices(model_params, filter_params)
+    @test minimum(mat_off.Lambda) > 0.0
+    ParticleDA.calculate_mean_height!(mat_on.mean, height, mat_off, obs, stations, model_params, filter_params)
+    @test all(isfinite.(mat_on.mean))
+
+    rng = Random.MersenneTwister(seed)
+    ParticleDA.sample_height_proposal!(height, mat_off, mat_on, obs, stations, model_params, filter_params, rng)
+    @test all(isfinite.(mat_on.samples))
+
+end
+
+@testset "Optimal Filter validation" begin
+
+    seed = 123
+    Random.seed!(seed)
+    rng = Random.MersenneTwister(seed)
+
+    include(joinpath(@__DIR__,"optimal_filter_validation.jl"));
+
+    params_dict = YAML.load_file(joinpath(@__DIR__, "optimal_filter_validation.yaml"))
+    filter_params = ParticleDA.get_params(FilterParameters, params_dict["filter"])
+    model_params = ParticleDA.get_params(ModelParameters, params_dict["model"]["llw2d"])
+
+    stations = (ist = st.st_ij[:,1].+1, jst = st.st_ij[:,2].+1)
+
+    h(x,y) = sin(x)^2 + cos(y)^2
+    height = zeros(model_params.nx, model_params.ny, filter_params.nprt)
+    x = (1:model_params.nx) .* 2 * pi / model_params.nx
+    y = (1:model_params.ny) .* 4 * pi / model_params.ny
+    for i = 1:filter_params.nprt
+        height[:,:,i] = h.(x',y)
```

Comment:
This becomes in-place :wink: 
```suggestion
        height[:,:,i] .= h.(x',y)
```